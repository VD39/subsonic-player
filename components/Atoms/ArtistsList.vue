<script setup lang="ts">
defineProps<{
  artists: BaseArtist[];
}>();
</script>

<template>
  <ul class="commaList">
    <li
      v-for="artist in artists"
      :key="artist.id"
      data-test-id="artist-list-item"
    >
      <NuxtLink :to="`/artist/${artist.id}`" class="link">
        {{ artist.name }}
      </NuxtLink>
    </li>
  </ul>
</template>
