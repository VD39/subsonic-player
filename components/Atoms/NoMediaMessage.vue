<script setup lang="ts">
defineProps<{
  icon: string;
  message: string;
}>();
</script>

<template>
  <div :class="$style.noMediaMessage">
    <div :class="['column', 'centerItems', $style.content]">
      <component :is="icon" :size="50" />

      <p>{{ message }}</p>
    </div>
  </div>
</template>

<style module>
.noMediaMessage {
  position: relative;
  height: var(--width-height-100);
}

.content {
  position: absolute;
  inset: 50% auto auto 50%;
  transform: translate(-50%, -50%);
}
</style>
