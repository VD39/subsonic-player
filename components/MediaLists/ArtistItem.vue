<script setup lang="ts">
import ImageLink from '@/components/Image/ImageLink.vue';

const props = defineProps<{
  artist: Artist;
}>();

const toLink = `/artist/${props.artist.id}`;
</script>

<template>
  <article>
    <ImageLink :to="toLink" :title="artist.name" :image="artist.image" />

    <p class="clamp clamp2">
      <NuxtLink :to="toLink" :title="`Go to artist ${artist.name}`">
        {{ artist.name }}
      </NuxtLink>
    </p>
  </article>
</template>
