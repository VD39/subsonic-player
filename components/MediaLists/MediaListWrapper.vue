<script setup lang="ts">
withDefaults(
  defineProps<{
    rows?: string;
  }>(),
  {
    rows: '5',
  },
);
</script>

<template>
  <section :class="$style.mediaListWrapper" :style="{ '--loop-rows': rows }">
    <slot />
  </section>
</template>

<style module>
.mediaListWrapper {
  --loop-cols: calc(
    (var(--loop-max-width) - (var(--loop-columns) - 1) * var(--loop-grid-gap)) /
      var(--loop-columns)
  );

  position: relative;
  display: grid;
  grid-template-columns: repeat(auto-fill, minmax(var(--loop-cols), 1fr));
  gap: var(--loop-grid-gap);
  list-style: none;

  @media (width >= 420px) {
    --loop-columns: max(var(--loop-rows) - 3, 2);
  }

  @media (width >= 775px) {
    --loop-columns: max(var(--loop-rows) - 2, 2);
  }

  @media (width >= 1000px) {
    --loop-columns: max(var(--loop-rows) - 1, 2);
  }

  @media (width >= 1260px) {
    --loop-columns: var(--loop-rows);
  }
}
</style>
