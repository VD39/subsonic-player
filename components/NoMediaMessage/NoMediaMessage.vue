<script setup lang="ts">
defineProps<{
  icon: string;
  message: string;
}>();
</script>

<template>
  <div :class="$style.noMediaMessage">
    <div :class="$style.content">
      <component :is="icon" :size="50" />

      <p>{{ message }}</p>
    </div>
  </div>
</template>

<style module>
.noMediaMessage {
  position: relative;
  height: 100%;
}

.content {
  @mixin align-center;

  position: absolute;
  inset: 50% auto auto 50%;
  flex-direction: column;
  transform: translate(-50%, -50%);
}
</style>
