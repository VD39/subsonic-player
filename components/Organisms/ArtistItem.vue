<script setup lang="ts">
import ImageLink from '@/components/Organisms/ImageLink.vue';

const props = defineProps<{
  artist: Artist | SimilarArtist;
}>();

const toLink = `/artist/${props.artist.id}`;
</script>

<template>
  <article>
    <ImageLink
      :to="toLink"
      :title="`Go to artist ${artist.name}`"
      :image="artist.image"
      class="mBS"
    />

    <p class="mBXS strong smallFont clamp2">
      <NuxtLink
        :to="toLink"
        class="link"
        :aria-label="`Go to artist ${artist.name}`"
      >
        {{ artist.name }}
      </NuxtLink>
    </p>
  </article>
</template>
