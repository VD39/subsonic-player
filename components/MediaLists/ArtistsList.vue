<script setup lang="ts">
import NoMediaMessage from '@/components/NoMediaMessage/NoMediaMessage.vue';
import MediaListWrapper from './MediaListWrapper.vue';
import ArtistItem from './ArtistItem.vue';

defineProps<{
  artists: Artist[];
  rows?: string;
}>();
</script>

<template>
  <MediaListWrapper v-if="artists.length" :row="rows">
    <ArtistItem v-for="artist in artists" :key="artist.id" :artist="artist" />
  </MediaListWrapper>
  <NoMediaMessage v-else icon="PhUsersThree" message="No artists found." />
</template>
