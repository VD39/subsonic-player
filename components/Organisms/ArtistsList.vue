<script setup lang="ts">
import MediaListWrapper from '@/components/Atoms/MediaListWrapper.vue';
import NoMediaMessage from '@/components/Atoms/NoMediaMessage.vue';
import ArtistItem from '@/components/Organisms/ArtistItem.vue';

defineProps<{
  artists: Artist[] | SimilarArtist[];
}>();
</script>

<template>
  <MediaListWrapper v-if="artists.length">
    <ArtistItem v-for="artist in artists" :key="artist.id" :artist />
  </MediaListWrapper>

  <NoMediaMessage
    v-else
    :icon="IMAGE_DEFAULT_BY_TYPE.artist"
    message="No artists found."
  />
</template>
