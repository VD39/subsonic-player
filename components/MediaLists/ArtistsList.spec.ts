import type { VueWrapper } from '@vue/test-utils';
import { mount } from '@vue/test-utils';
import { getFormattedArtistsMock } from '@/test/helpers';
import NoMediaMessage from '@/components/NoMediaMessage/NoMediaMessage.vue';
import MediaListWrapper from './MediaListWrapper.vue';
import ArtistItem from './ArtistItem.vue';
import ArtistsList from './ArtistsList.vue';

function factory(props = {}) {
  return mount(ArtistsList, {
    props: {
      artists: [],
      ...props,
    },
  });
}

describe('ArtistsList', () => {
  let wrapper: VueWrapper;

  beforeAll(() => {
    wrapper = factory();
  });

  it('matches the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  describe('when artists prop is an empty array', () => {
    it('does not show the MediaListWrapper component', () => {
      expect(wrapper.findComponent(MediaListWrapper).exists()).toBe(false);
    });

    it('shows the NoMediaMessage component', () => {
      expect(wrapper.findComponent(NoMediaMessage).exists()).toBe(true);
    });
  });

  describe('when artists prop is not an empty array', () => {
    beforeAll(() => {
      wrapper = factory({
        artists: getFormattedArtistsMock(5),
      });
    });

    it('matches the snapshot', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    it('shows the MediaListWrapper component', () => {
      expect(wrapper.findComponent(MediaListWrapper).exists()).toBe(true);
    });

    it('shows the correct number of ArtistItem components', () => {
      expect(wrapper.findAllComponents(ArtistItem).length).toBe(5);
    });

    it('does not show the NoMediaMessage component', () => {
      expect(wrapper.findComponent(NoMediaMessage).exists()).toBe(false);
    });
  });
});
