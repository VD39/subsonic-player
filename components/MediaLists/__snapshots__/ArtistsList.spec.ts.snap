// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ArtistsList > matches the snapshot 1`] = `
"<div class="noMediaMessage">
  <div class="content">
    <phusersthree size="50"></phusersthree>
    <p>No artists found.</p>
  </div>
</div>"
`;

exports[`ArtistsList > when artists prop is not an empty array > matches the snapshot 1`] = `
"<section class="mediaListWrapper" style="--loop-rows: 5;">
  <article>
    <figure class="figure"><a title="Go to artist-0">
        <div class="preloadImage"><span class="loader"><span class="visually-hidden">Loading image</span></span><img class="image" src="coverArt" alt="artist-0" style="display: none;"></div>
      </a></figure>
    <p class="clamp clamp2"><a title="Go to artist artist-0">artist-0</a></p>
  </article>
  <article>
    <figure class="figure"><a title="Go to artist-1">
        <div class="preloadImage"><span class="loader"><span class="visually-hidden">Loading image</span></span><img class="image" src="coverArt" alt="artist-1" style="display: none;"></div>
      </a></figure>
    <p class="clamp clamp2"><a title="Go to artist artist-1">artist-1</a></p>
  </article>
  <article>
    <figure class="figure"><a title="Go to artist-2">
        <div class="preloadImage"><span class="loader"><span class="visually-hidden">Loading image</span></span><img class="image" src="coverArt" alt="artist-2" style="display: none;"></div>
      </a></figure>
    <p class="clamp clamp2"><a title="Go to artist artist-2">artist-2</a></p>
  </article>
  <article>
    <figure class="figure"><a title="Go to artist-3">
        <div class="preloadImage"><span class="loader"><span class="visually-hidden">Loading image</span></span><img class="image" src="coverArt" alt="artist-3" style="display: none;"></div>
      </a></figure>
    <p class="clamp clamp2"><a title="Go to artist artist-3">artist-3</a></p>
  </article>
  <article>
    <figure class="figure"><a title="Go to artist-4">
        <div class="preloadImage"><span class="loader"><span class="visually-hidden">Loading image</span></span><img class="image" src="coverArt" alt="artist-4" style="display: none;"></div>
      </a></figure>
    <p class="clamp clamp2"><a title="Go to artist artist-4">artist-4</a></p>
  </article>
</section>"
`;
